src/daft-core/src/array/ops/cast.rs

```diff
@@ -593,693 +592,23 @@ impl FileArray {
-                    Err(DaftError::ValueError(format!(
-                        "We can only convert numeric python types to FixedSizeList, got {}",
-                        child_dtype,
-                    )))
-                }
-            }
-            DataType::Struct(_) => {
-                // TODO: This is not fully operational since we can't guarantee that our Series.from_pylist
-                // implementation (using PyArrow) will cast the struct field arrays correctly.
-                pycast_then_arrowcast!(self, dtype, "dict")
-            }
-            DataType::Embedding(..) => {
-                let result = self.cast(&dtype.to_physical())?;
-                let embedding_array = EmbeddingArray::new(
-                    Field::new(self.name(), dtype.clone()),
-                    result.fixed_size_list()?.clone(),
-                );
-                Ok(embedding_array.into_series())
-            }
-            DataType::Image(mode) => {
-                let inner_dtype = mode.map_or(DataType::UInt8, |m| m.get_dtype());
-                with_match_numeric_daft_types!(inner_dtype, |$T| {
-                    type Tgt = <$T as DaftNumericType>::Native;
-                    pyo3::Python::with_gil(|py| {
-                        let result = extract_python_like_to_image_array::<Tgt>(py, self, dtype, &inner_dtype, *mode)?;
-                        Ok(result.into_series())
-                    })
-                })
-            }
-            DataType::FixedShapeImage(..) => {
-                let result = self.cast(&dtype.to_physical())?;
-                let image_array = FixedShapeImageArray::new(
-                    Field::new(self.name(), dtype.clone()),
-                    result.fixed_size_list()?.clone(),
-                );
-                Ok(image_array.into_series())
-            }
-            DataType::Tensor(inner_dtype) => {
-                with_match_numeric_daft_types!(**inner_dtype, |$T| {
-                    type Tgt = <$T as DaftNumericType>::Native;
-                    pyo3::Python::with_gil(|py| {
-                        let result = extract_python_like_to_tensor_array::<Tgt>(py, self, dtype, &inner_dtype)?;
-                        Ok(result.into_series())
-                    })
-                })
-            }
-            DataType::FixedShapeTensor(..) => {
-                let result = self.cast(&dtype.to_physical())?;
-                let tensor_array = FixedShapeTensorArray::new(
-                    Field::new(self.name(), dtype.clone()),
-                    result.fixed_size_list()?.clone(),
-                );
-                Ok(tensor_array.into_series())
-            }
-            // TODO: Add implementations for these types
-            // DataType::Timestamp(_, _) => $self.timestamp().unwrap().$method($($args),*),
-            dt => unimplemented!("dtype {:?} not supported", dt),
-        }
+        Ok(Series::from_literals(literals)?
+            .cast(dtype)?
```

Comment:
Can we update `Series::from_literals` to instead consume an iterator? that way we don't need to `.collect()` just to immediately pass it in to `from_literals`. 

instead you should just be able to collect into a series `.collect::<PyResult<Series>>()`

Reply:
Hm I'm not sure how to implement `Series::from_literals` with an iterator input since it would need to consume the iterator twice, first to get the data type at the start and second to actually turn it into a series.

Also it would not be very useful here because the literal iterator would have item type `PyResult<Literal>`, so we do still need to collect it to turn it into an iterator of literals.

Reply:
Since this is now in a very common path, can we atleast add a `TODO` on looking into ways to optimize this. I think we can do better than 2-3x iterations on every cast. 
(1) pythonarray -> vec<literal>
(2) vec<literal> -> Series
(3) series.cast

Reply:
Will do. I might be wrong here but I think with the better default type conversions that we will have with this PR, I don't see people using `.cast` on a Python column all that often anymore. Most of the time, they can probably just specify the right data type in the UDF definition and we avoid the cast altogether.

Reply:
Before, we relied on cast to go from Python to tensors or images, but that is no longer the case with these changes.